ethexe/compute/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2024-2025 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Sequencer for ethexe.
+
+use anyhow::{anyhow, Result};
+use ethexe_common::{
+    db::{BlockMetaStorage, CodesStorage},
+    events::{BlockEvent, BlockRequestEvent, RouterRequestEvent},
+    gear::{BlockCommitment, CodeCommitment, StateTransition},
+    BlockData,
+};
+use ethexe_db::{CodeInfo, Database};
+use ethexe_observer::Query;
+use ethexe_processor::{LocalOutcome, Processor};
+use futures::{future::BoxFuture, stream::FusedStream, FutureExt, Stream};
+use gprimitives::{CodeId, H256};
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::task::JoinSet;
+
+#[derive(Debug)]
+pub struct BlockProcessed {
+    pub chain_head: H256,
+    // TODO (gsobol): remove commitments, this must be handled by validator if needed
+    pub commitments: Vec<BlockCommitment>,
+}
+
+#[derive(Debug)]
+pub enum ComputeEvent {
+    BlockProcessed(BlockProcessed),
+    CodeProcessed(CodeCommitment),
+}
+
+// TODO (gsobol): add state monitoring in prometheus
+// TODO (gsobol): append off-chain transactions handling
+// TODO (gsobol) asap: add tests for compute service
+pub struct ComputeService {
+    db: Database,
+    processor: Processor,
+    query: Query,
+    blocks_queue: VecDeque<BlockData>,
+    process_block: Option<BoxFuture<'static, Result<BlockProcessed>>>,
+    process_codes: JoinSet<Result<CodeCommitment>>,
```

Comment:
pls use FuturesUnordered

Reply:
Where?

Reply:
So, I think changing process_codes: JoinSet<Result<CodeCommitment>> to FuturesUnordered is useless here.